src/aof.c

```diff
@@ -548,12 +557,101 @@ int persistAofManifest(aofManifest *am) {
     }
 
     sds amstr = getAofManifestAsString(am);
-    int ret = writeAofManifestFile(amstr);
+    int ret = writeAofManifestFile(server.aof_filename, amstr);
     sdsfree(amstr);
     if (ret == C_OK) am->dirty = 0;
     return ret;
 }
 
+/* Called in `loadAppendOnlyFiles` when we upgrade from a old version redis.
+ * 
+ * 1) Check if the UPGRADE_TEMP_DIR_NAME directory already exists.
+ *    1a) If it exists, then check if 'server.aof_filename' file exists in this 
+ *        directory, If it exists, then continue to execute rename and return.
+ *    1b) If it exists and the 'server.aof_filename' file not exists in this 
+ *        directory, then we directly delete UPGRADE_TEMP_DIR_NAME directory.
+ * 2) Create UPGRADE_TEMP_DIR_NAME directory.
+ * 3) Use 'server.aof_filename' to construct a BASE type aofInfo and add it to 
+ *    aofManifest, then persist the manifest file to UPGRADE_TEMP_DIR_NAME directory.
+ * 4) Move the old AOF file (server.aof_filename) to UPGRADE_TEMP_DIR_NAME directory.
+ * 5) rename UPGRADE_TEMP_DIR_NAME to server.aof_filename.
+ */
+int aofUpgradePrepare(aofManifest *am) {
+    sds aof_filepath = makePath(UPGRADE_TEMP_DIR_NAME, server.aof_filename);
```

Comment:
I think the temp dir must be based on the server.aof_filename pattern, and not a constant one.
otherwise, if there are two servers in the same dir, their temp upgrade dir clash.